File: templating/rule_contributions/src/main/python/jsonteng_contribs/rules/ipv4_subnet_rule.py
# Copyright 2019 VMware, Inc.
# SPDX-License-Indentifier: Apache-2.0

import ipaddress

from jsonteng.exception import TemplateEngineException
from jsonteng.rules.rule_base import RuleBase


class Ipv4SubnetRule(RuleBase):
    """
    Divide a network into N number of subnets. Return i_th of the subnet.
    """
    name = "ipv4-subnet"

    def __init__(self, rule_resolver):
        """
        Construct this rule.
        :param rule_resolver: Rule resolver
        :type rule_resolver: 'RuleResolver'
        """
        super().__init__(rule_resolver)
        self._element_resolver = rule_resolver.get_element_resolver()

    def process(self, rule_tokens, binding_data_list):
        """
        Process this rule.
        :param rule_tokens: Rule arguments.
        :type rule_tokens: 'list'
        :param binding_data_list: Binding data used during the processing.
        :type binding_data_list: 'list'
        :return: JSON object
        :rtype: JSON object
        """
        token_count = len(rule_tokens)
        if token_count < 3:
            raise TemplateEngineException(
                "Rule \"{}\" needs at least 3 parameter."
                " Parameters given {}".format(Ipv4SubnetRule.name, rule_tokens))
        network = ipaddress.ip_network(self._element_resolver.resolve(
            rule_tokens[0], binding_data_list))
        subnet_count = int(self._element_resolver.resolve(
            rule_tokens[1], binding_data_list))
        subnet_index = int(self._element_resolver.resolve(
            rule_tokens[2], binding_data_list))
        base2_exp = 0
        count = int(subnet_count) - 1
        while count & 1:
            base2_exp += 1
            count = count >> 1
        if count:
            raise TemplateEngineException(
                "Subnet count must be multiple of 2s."
                " {} is given".format(subnet_count))
        if base2_exp == 0:
            return rule_tokens[0]
        subnet_prefix = network.prefixlen + base2_exp
        subnet_address = ipaddress.ip_address(
            int(network.network_address)
            + (subnet_index << (32 - subnet_prefix)))
        return str(subnet_address) + '/' + str(subnet_prefix)

File: templating/rule_contributions/src/main/java/com/vmware/jsonteng/ext/rules/net/Ipv4SubnetRule.java
package com.vmware.jsonteng.ext.rules.net;

import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.nio.ByteBuffer;
import java.util.List;
import java.util.Map;

import com.vmware.jsonteng.ElementResolver;
import com.vmware.jsonteng.RuleResolver;
import com.vmware.jsonteng.TemplateEngineException;
import com.vmware.jsonteng.rules.RuleBase;

public class Ipv4SubnetRule extends RuleBase {
    public static final String name = "ipv4-subnet";

    private final ElementResolver elementResolver;

    public Ipv4SubnetRule(RuleResolver ruleResolver) {
        super(ruleResolver);
        elementResolver = ruleResolver.getElementResolver();
    }

    @Override
    public Object process(List<?> ruleTokens, List<Map<String, ?>> bindingDataList) throws TemplateEngineException {
        int tokenCount = ruleTokens.size();
        if (tokenCount < 3) {
            throw new TemplateEngineException(
                    String.format("Rule \"%s\" needs at least 3 parameters. Parameters given %s", name, ruleTokens));
        }
        String network = (String) this.elementResolver.resolve(ruleTokens.get(0), bindingDataList);
        int subnetCount = Integer.parseInt(this.elementResolver.resolve(
                ruleTokens.get(1), bindingDataList).toString());
        int subnetIndex = Integer.parseInt(this.elementResolver.resolve(
                ruleTokens.get(2), bindingDataList).toString());
        int base2Exp = 0;
        int count = subnetCount - 1;
        while ((count & 0x1) != 0) {
            count = count >>> 1;
            base2Exp += 1;
        }
        if (count != 0) {
            throw new TemplateEngineException(
                    String.format("Subnet count must be multiple of 2s. %d is given", subnetCount));
        }
        if (base2Exp == 0) {
            return network;
        }
        String[] parts = network.split("/");
        Inet4Address networkAddress;
        try {
            networkAddress = (Inet4Address) InetAddress.getByName(parts[0]);
        } catch (UnknownHostException e) {
            throw new TemplateEngineException("Unable to parse ip address " + parts[0], e);
        }
        int networkPrefix = Integer.parseInt(parts[1]);
        int subnetPrefix = networkPrefix + base2Exp;
        ByteBuffer fourBytes = ByteBuffer.wrap(new byte[4]);
        fourBytes.put(networkAddress.getAddress());
        fourBytes.rewind();
        int networkBytes = fourBytes.getInt();
        int subnetBytes = networkBytes + (subnetIndex << (32 - subnetPrefix));
        fourBytes.rewind();
        fourBytes.putInt(subnetBytes);
        Inet4Address subnetAddress;
        try {
            subnetAddress = (Inet4Address) InetAddress.getByAddress(fourBytes.array());
        } catch (UnknownHostException e) {
            throw new TemplateEngineException("Unable to partition subnet " + fourBytes.toString(), e);
        }
        return subnetAddress.getHostAddress() + '/' + Integer.toString(subnetPrefix);
    }
}
